pageserver/src/tenant/timeline/compaction.rs

```diff
@@ -668,10 +668,31 @@ impl Timeline {
 
         // Define partitioning schema if needed
 
-        // FIXME: the match should only cover repartitioning, not the next steps
-        let (partition_count, has_pending_tasks) = match self
+        // 1. L0 Compact
+        let fully_compacted = {
+            let timer = self.metrics.compact_time_histo.start_timer();
+            let fully_compacted = self
+                .compact_level0(
+                    target_file_size,
+                    options.flags.contains(CompactFlags::ForceL0Compaction),
+                    ctx,
+                )
+                .await?;
+            timer.stop_and_record();
+            fully_compacted
+        };
+
+        if !fully_compacted {
+            // Yield and do not do any other kind of compaction. True means
+            // that we have pending L0 compaction tasks and the compaction scheduler
+            // will prioritize compacting this tenant/timeline again.
+            return Ok(true);
+        }
```

Comment:
There is a risk here that we'll starve image layer compaction if the L0 growth rate is too high. I believe we said that this wasn't critical, since reads would terminate at the last page image anyway (at the checkpoint interval) -- that should be fine then, but let's keep an eye on this when it's deployed.

Reply:
Good point.